### PR TITLE
Feature/role based access controll 3

### LINKED ISSUE #14: Store Custom Roles for user authorities and restrictions in a database instead of enum
Storing the authorities and restrictions in a database is more flexible and dynamic.
@muttaqin1

### PR DESCRIPTION
**Pull request by muttaqin1**
resolved #14 
- Dynamic roles 
- Role controller

I have removed roles parameter in restrict to middleware. The middleware now validates user using actions.
there will be some ACTIONS related to superAdmin
 such as:
```js
const Actions= {
REMOVE_ADMIN:"REMOVE_ADMIN",
MANAGE_ROLES:"MANAGE_ROLES"
}
```
now the routes for super admin will use this ACTIONS. User with the required authorities can access those routes.
please review my changes.
@bellaabdelouahab

### CODE REVIEW

File: backend-app/utils/authorization/role/Role.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        type: role.name,
+        authorities: role.authorities,
+        restrictions: role.restrictions,
+      },
+      new AppError(500, 'fail', 'Internal server error!')
+    );
+  }
+}
+
+module.exports = Role;
```
Comment (bellaabdelouahab) [inline backend-app/utils/authorization/role/Role.js:145]: at this stage we will need to start code testing, to verify that all these functions will always work, 
- @Khalid1G , you told me you want to start code testing right 😁.

**Review by bellaabdelouahab — APPROVED** [review]
we will definitely need some code testing for this

File: backend-app/utils/authorization/role/Role.js (hunk 1 above)
Comment (Khalid1G) [reply backend-app/utils/authorization/role/Role.js:145]: both of us, @bellaabdelouahab

**Comment by muttaqin1** [comment]
> we will definitely need some code testing for this

Here comes the boring part😁. btw, can you provide all the Actions?

**Comment by muttaqin1** [comment]
> we will definitely need some code testing for this  

yeah but testing is so boring 😁,
 The createAdminUser function is not getting **authorities** and **restrictions**.
 can we keep it in the **authorization** dir?.

File: backend-app/utils/authorization/role/Role.js (hunk 1 above)
Comment (bellaabdelouahab) [reply backend-app/utils/authorization/role/Role.js:145]: sure.

**Comment by bellaabdelouahab** [comment]
> > we will definitely need some code testing for this
> 
> yeah but testing is so boring 😁, The createAdminUser function is not getting **authorities** and **restrictions**. can we keep it in the **authorization** dir?.

i dint understand you

**Comment by bellaabdelouahab** [comment]
anyway we will keep you updated on testing @muttaqin1

**Comment by muttaqin1** [comment]
```js
const createAdminUser = async () => {
  try {
    const user = await User.findOne({ email: ADMIN_EMAIL });
    if (!user) {
      await User.create({
        name: 'Admin',
        email: ADMIN_EMAIL,
        password: ADMIN_PASSWORD,
        roles: ['SUPER_ADMIN'],
      });
      Logger.info('Admin user created successfully');
    }
  } catch (err) {
    Logger.error(err);
  }
};

```
This function is only getting the role but  not getting the authorities and restrictions.

```js
const createAdminUser = async () => {
  try {
    const user = await User.findOne({ email: ADMIN_EMAIL });
    if (!user) {
      await User.create({
        name: 'Admin',
        email: ADMIN_EMAIL,
        password: ADMIN_PASSWORD,
        roles: ['SUPER_ADMIN'],
//authorities: [], missing!
//restrictions:[] missing!
      });
      Logger.info('Admin user created successfully');
    }
  } catch (err) {
    Logger.error(err);
  }
};
```

**Pull request merged**